lusSTR/snps.py

```diff
@@ -0,0 +1,329 @@
+        ]
+        data_df.append(row_tmp)
+    data_final = pd.DataFrame(data_df, columns=[
+        'SampleID', 'Project', 'Analysis', 'SNP', 'Reads', 'Forward_Strand_Allele', 'UAS_Allele',
+        'Type', 'Issues'
+    ])
+    return data_final
+
+
+def compile_row_of_snp_data(infile, snp, table_loc, type, name, analysis):
+    '''
+    This function is necessary to account for the two sets of SNPs reported from the same
+    sequence amplicon. Sequences labeled as mh16-MC1RB and mh16-MC1RC contain 3 and 6 SNPs,
+    respectively. This function reports out each SNP from the sequence amplicon as individual
+    rows and calls another function to compile data on each SNP.
+    '''
+    snp_df = []
+    if 'mh16' in snp:
+        locus_data = snps_within_loci[snp]
+        for k in range(0, len(locus_data['SNPs'])):
+            snp_id = locus_data['SNPs'][k]
+            row_tmp = collect_snp_info(infile, snp_id, table_loc, type, name, analysis)
+            if row_tmp is not None:
+                snp_df.append(row_tmp)
+    else:
+        row_tmp = collect_snp_info(infile, snp, table_loc, type, name, analysis)
+        if row_tmp is not None:
+            snp_df.append(row_tmp)
+    final_snp_df = pd.DataFrame(snp_df)
+    return final_snp_df
+
+
+def snp_call_exception(seq, expected_size, metadata, base):
+    '''
+    This function accounts for insertions and deletions in sequences to identify the correct base
+     for the SNP. If the identified allele is still not one of the expected alleles, the sequence
+     will be flagged appropriately.
+    '''
+    new_size = len(seq) + expected_size
+    new_base_call = seq[new_size]
+    if new_base_call in metadata['Alleles']:
+        flag = (
+            'Sequence length different than expected (check for indels); allele position adjusted'
+        )
+        return new_base_call, flag
+    else:
+        flag = (
+            'Allele call does not match expected allele! Check for indels '
+            '(does not match expected sequence length)'
+        )
+        return base, flag
+
+
+def collect_snp_info(infile, snpid, j, type, name, analysis):
+    '''
+    This function compiles allele calls, reads, reverse complements allele call if necessary to
+    match how the UAS reports the allele, and any flags associated with the allele call. The flags
+    indicate potential issues, including an unexpected allele call (not one of two expected
+    alleles for the SNP) or unexpected length of the sequence amplicon which could result in an
+    incorrect allele call. This function also determines if the SNP should be included in the
+    final table based on the specified SNP type from the CLI.
+    '''
+    if snpid == 'N29insA':
+        snpid = 'rs312262906_N29insA'
+    metadata = snp_marker_data[snpid]
+    snp_type = metadata['Type']
+    seq = infile.iloc[j, 2]
+    expected_alleles = metadata['Alleles']
+    snp_loc = metadata['Coord']
+    snp_call = seq[snp_loc]
```

Comment:
So these two lines are the primary logic for determining the SNP allele, right? Everything else seems to be handling special cases or putative and unexpected indels.

Reply:
Yup!